### PR TITLE
Separate compareFunctions from alive-tv.cpp

### PR DESCRIPTION
**Pull request by Stefan-Mada**
`compareFunctions` is a very useful function for comparing LLVM functions for refinement, but it is currently inside `alive-tv.cpp`. This makes it impossible to include the code in other projects due to the main function, so I have separated out its functionality into `function_verify.cpp` and `function_verify.h`. I also added an additional function `initialize` which sets `opt_quiet`, `opt_always_verify`, `out`, etc, in order to keep from polluting the namespace.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -187,6 +187,7 @@ if (BUILD_LLVM_UTILS OR BUILD_TV)
     "llvm_util/llvm_optimizer.cpp"
     "llvm_util/llvm2alive.cpp"
     "llvm_util/utils.cpp"
+    "llvm_util/function_verify.cpp"
```
Comment (nunoplopes) [inline CMakeLists.txt:190]: please keep it sorted

File: tools/alive-tv.cpp [hunk 2]
```diff
@@ -87,168 +82,7 @@ std::unique_ptr<llvm::Module> openInputFile(llvm::LLVMContext &Context,
...
-      if (!opt_quiet)
-        *out << r.errs << endl;
-      return false;
-    }
-  }
-  return true;
-}
-
-llvm::Function *findFunction(llvm::Module &M, const string &FName) {
+llvm::Function *findFunction(llvm::Module &M, const std::string &FName) {
```
Comment (nunoplopes) [inline tools/alive-tv.cpp:85]: no std:: here pls

File: tools/alive-tv.cpp [hunk 3]
```diff
@@ -260,7 +94,6 @@ llvm::Function *findFunction(llvm::Module &M, const string &FName) {
 }
 }
 
-
```
Comment (nunoplopes) [inline tools/alive-tv.cpp:263]: please leave my whitespace alone..

File: llvm_util/function_verify.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+bool compareFunctions(llvm::Function &F1, llvm::Function &F2,
+                      llvm::TargetLibraryInfoWrapperPass &TLI);
+
+
+void initialize_verifier(bool set_opt_quiet, bool set_opt_always_verify,
+                         bool set_opt_print_dot, bool set_opt_bidirectional,
+                         std::ostream *set_out);
+
+}
```
Comment (nunoplopes) [inline llvm_util/function_verify.h:35]: missing EOL

File: llvm_util/function_verify.h [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+// Distributed under the MIT license that can be found in the LICENSE file.
+
+#include "cache/cache.h"
+#include "llvm_util/llvm2alive.h"
+#include "smt/smt.h"
+#include "tools/transform.h"
+
+#include "llvm/Analysis/TargetLibraryInfo.h"
+
+#include <iostream>
```
Comment (nunoplopes) [inline llvm_util/function_verify.h:13]: no iostream in headers. use ostream isntead

File: llvm_util/function_verify.cpp [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+using namespace std;
+using namespace llvm_util;
+
+namespace {
+bool opt_quiet = false;
+bool opt_always_verify = false;
+bool opt_print_dot = false;
+bool opt_bidirectional = false;
+
+ostream* out = &cout;
```
Comment (nunoplopes) [inline llvm_util/function_verify.cpp:17]: *out, not * out

File: llvm_util/function_verify.h [hunk 7]
```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+// Copyright (c) 2022-present The Alive2 Authors.
+// Distributed under the MIT license that can be found in the LICENSE file.
+
+#include "cache/cache.h"
+#include "llvm_util/llvm2alive.h"
```
Comment (nunoplopes) [inline llvm_util/function_verify.h:7]: this header doesn't belong here I think

File: llvm_util/function_verify.h [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+#include "smt/smt.h"
+#include "tools/transform.h"
```
Comment (nunoplopes) [inline llvm_util/function_verify.h:9]: this header doesn't belong here

File: llvm_util/function_verify.h [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
```
Comment (nunoplopes) [inline llvm_util/function_verify.h:10]: no space here

File: llvm_util/function_verify.h [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+extern unsigned num_unsound;
+extern unsigned num_failed;
+extern unsigned num_errors;
+
+
+bool compareFunctions(llvm::Function &F1, llvm::Function &F2,
+                      llvm::TargetLibraryInfoWrapperPass &TLI);
+
+
+void initialize_verifier(bool set_opt_quiet, bool set_opt_always_verify,
```
Comment (nunoplopes) [inline llvm_util/function_verify.h:31]: can be called initialize only since it's in a namespace.

File: llvm_util/function_verify.h [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+#include "llvm_util/llvm2alive.h"
+#include "smt/smt.h"
+#include "tools/transform.h"
+
+#include "llvm/Analysis/TargetLibraryInfo.h"
+
+#include <iostream>
+#include <sstream>
+
+namespace verifier {
```
Comment (nunoplopes) [inline llvm_util/function_verify.h:16]: llvm_util::verifier

**Comment by Stefan-Mada** [comment]
Hey Nuno,
I implemented all the changes you wanted! I see that the checks fail, but it's the same 3 unexpected failures as on the main branch, so it doesn't appear to be an issue with this pull request specifically. Please take a look whenever it's convenient.

**Comment by nunoplopes** [comment]
The function is now separated through another PR. I'll merge anything from this PR if relevant. If something else is missing for you, please open another PR/issue.
Thank you!

**Pull request closed without merging**